device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/exceptions/IotHubClientException.java

```diff
@@ -0,0 +1,180 @@
+        }
+    }
+
+    public static IotHubClientException fromStatusCode(IotHubStatusCode statusCode)
+    {
+        return fromStatusCode(statusCode, "");
+    }
+
+    public static IotHubClientException fromStatusCode(IotHubStatusCode statusCode, String errorMessage)
+    {
+        boolean isRetryable = IotHubStatusCode.isRetryable(statusCode);
+        switch (statusCode)
+        {
+            case SERVER_BUSY:
+            case THROTTLED:
+                return new IotHubClientThrottledException(errorMessage, isRetryable);
+            case QUOTA_EXCEEDED:
+                return new IotHubClientQuotaExceededException(errorMessage, isRetryable);
+            case REQUEST_ENTITY_TOO_LARGE:
+                return new IotHubClientMessageTooLargeException(errorMessage, isRetryable);
+            case PRECONDITION_FAILED:
+                return new IotHubClientPreconditionFailedException(errorMessage, isRetryable);
+            case UNAUTHORIZED:
+                return new IotHubClientUnauthorizedException(errorMessage, isRetryable);
+            case HUB_OR_DEVICE_ID_NOT_FOUND:
+                return new IotHubClientNotFoundException(errorMessage, isRetryable);
+            case BAD_FORMAT:
+                return new IotHubClientBadFormatException(errorMessage, isRetryable);
+            case INTERNAL_SERVER_ERROR:
+                return new IotHubClientInternalServerErrorException(errorMessage, isRetryable);
+            case MESSAGE_EXPIRED:
+                return new IotHubClientMessageExpiredException(errorMessage, isRetryable);
+            case ERROR:
+                return new IotHubClientException(errorMessage, isRetryable);
+            case MESSAGE_CANCELLED_ONCLOSE:
+                return new IotHubClientMessageCancelledOnCloseException(errorMessage, isRetryable);
+            case OK:
+                return null;
+        }
+        return null;
+    }
+
+    public IotHubClientException()
+    {
+        this(true);
+    }
+
+    public IotHubClientException(boolean isRetryable)
+    {
+        super();
+        this.isRetryable = isRetryable;
+    }
+
+    public IotHubClientException(String message)
+    {
+        this(message, true);
+    }
+
+    public IotHubClientException(String message, boolean isRetryable)
+    {
+        super(message);
+        this.isRetryable = isRetryable;
+    }
+
+    public IotHubClientException(String message, Throwable cause)
+    {
+        this(message, cause, true);
+    }
+
+    public IotHubClientException(String message, Throwable cause, boolean isRetryable)
+    {
+        super(message, cause);
+        this.isRetryable = isRetryable;
+    }
+
+    public IotHubClientException(Throwable cause)
+    {
+        this(cause, true);
+    }
+
+    public IotHubClientException(Throwable cause, boolean isRetryable)
+    {
+        super(cause);
+        this.isRetryable = isRetryable;
+    }
+
+    @Getter
+    protected boolean isRetryable;
```

Comment:
Did we have this originally? how do you feel about calling this `isTransient`?

Reply:
We did not have this originally, no. I like isRetryable a bit better if only because "transient" might require some users to lookup its definition

Reply:
fair enough